### PR TITLE
i.atcorr: Fix uninitialized variable

### PR DESCRIPTION
**Pull request by ShubhamDesai**
**This pull request addresses uninitialized variable issue for all cpp files in imagery**

**Issue:**
i.atcorr/main.cpp:213:16: error: Uninitialized variable: rbitem.ti [uninitvar]
return rbitem;

**Changes Made:**
Initialized the RBitem struct to zero to ensure all members, including nested structures, are properly initialized.

### CODE REVIEW

**Comment by echoix** [comment]
There was only one file left?

**Comment by ShubhamDesai** [comment]
> There was only one file left?

yes in imagery folder there is only one cpp file left other files are c file. So now i am grouping folder wise and programming language wise

**Comment by echoix** [comment]
> > There was only one file left?
> 
> 
> 
> yes in imagery folder there is only one cpp file left other files are c file. So now i am grouping folder wise and programming language wise

Perfect!

**Comment by echoix** [comment]
See the new error here (and build failure) (ignore the garbage of the html file, it seems like unsynced output of parallel tasks):
```
<r.relief.html> is missing the descriptionmain.cpp: In member function ‘RBitem TICache::set_alt_vis(double, double)’:
main.cpp:207:34: error: missing initializer for member ‘RBitem::vis’ [-Werror=missing-field-initializers]
  207 |         struct RBitem rbitem = {0};
      |                                  ^
main.cpp:207:34: error: missing initializer for member ‘RBitem::ti’ [-Werror=missing-field-initializers]
cc1plus: all warnings being treated as errors
make[3]: *** [../../include/Make/Compile.make:35: OBJ.x86_64-pc-linux-gnu/main.o] Error 1
```

**Comment by ShubhamDesai** [comment]
> See the new error here (and build failure) (ignore the garbage of the html file, it seems like unsynced output of parallel tasks):
> 
> ```
> <r.relief.html> is missing the descriptionmain.cpp: In member function ‘RBitem TICache::set_alt_vis(double, double)’:
> main.cpp:207:34: error: missing initializer for member ‘RBitem::vis’ [-Werror=missing-field-initializers]
>   207 |         struct RBitem rbitem = {0};
>       |                                  ^
> main.cpp:207:34: error: missing initializer for member ‘RBitem::ti’ [-Werror=missing-field-initializers]
> cc1plus: all warnings being treated as errors
> make[3]: *** [../../include/Make/Compile.make:35: OBJ.x86_64-pc-linux-gnu/main.o] Error 1
> ```
> 
> https://github.com/OSGeo/grass/actions/runs/9880555579/job/27289369496?pr=4022#step:8:8216

let me check

**Review by echoix — DISMISSED** [review]

File: imagery/i.atcorr/main.cpp [hunk 1]
```diff
@@ -204,7 +204,19 @@ class TICache {
 private:
     struct RBitem set_alt_vis(double alt, double vis)
     {
-        struct RBitem rbitem;
+        struct RBitem rbitem = {0,
```
Comment (lbartoletti) [inline imagery/i.atcorr/main.cpp:207]: Instead of having a lot of 0, 0.0 could we have a function to initialize this struct?

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp:207]: The struct has two nested structs.
So i could just think of simple way in which it could be initialize.
No idea how it could be used with function

Comment (nilason) [reply imagery/i.atcorr/main.cpp:207]: > Instead of having a lot of 0, 0.0 could we have a function to initialize this struct?

This _is_ the function to initialize the struct.

A perhaps more clearer solution could be (diff against main):

```diff
diff --git a/imagery/i.atcorr/main.cpp b/imagery/i.atcorr/main.cpp
index 8b2a1779fa..fc37af791b 100644
--- a/imagery/i.atcorr/main.cpp
+++ b/imagery/i.atcorr/main.cpp
@@ -204,11 +204,22 @@ class TICache {
 private:
     struct RBitem set_alt_vis(double alt, double vis)
     {
-        struct RBitem rbitem;
-
         /* alt and vis must be in meters */
-        rbitem.alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5));
-        rbitem.vis = (int)(vis + 0.5);
+        struct RBitem rbitem = {
+            .alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5)),
+            .vis = (int)(vis + 0.5),
+            .ti = {
+                .asol = 0.0,
+                .ainr = {{0.0, 0.0, 0.0}, {0.0, 0.0, 0.0}},
+                .sb = 0.0,
+                .seb = 0.0,
+                .tgasm = 0.0,
+                .sutott = 0.0,
+                .sdtott = 0.0,
+                .sast = 0.0,
+                .srotot = 0.0,
+                .xmus = 0.0,
+            }};
 
         return rbitem;
     }
```

Comment (nilason) [reply imagery/i.atcorr/main.cpp:207]: This shorter version seem to satisfy cppcheck:
```diff
diff --git a/imagery/i.atcorr/main.cpp b/imagery/i.atcorr/main.cpp
index 8b2a1779fa..532ec58456 100644
--- a/imagery/i.atcorr/main.cpp
+++ b/imagery/i.atcorr/main.cpp
@@ -204,11 +204,11 @@ class TICache {
 private:
     struct RBitem set_alt_vis(double alt, double vis)
     {
-        struct RBitem rbitem;
-
         /* alt and vis must be in meters */
-        rbitem.alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5));
-        rbitem.vis = (int)(vis + 0.5);
+        struct RBitem rbitem = {
+            .alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5)),
+            .vis = (int)(vis + 0.5),
+        };
 
         return rbitem;
     }
```

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp]: i have done the recommended changes

Comment (lbartoletti) [reply imagery/i.atcorr/main.cpp]: Thanks :)

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp]: > Thanks :)



> This shorter version seem to satisfy cppcheck:
> 
> ```diff
> diff --git a/imagery/i.atcorr/main.cpp b/imagery/i.atcorr/main.cpp
> index 8b2a1779fa..532ec58456 100644
> --- a/imagery/i.atcorr/main.cpp
> +++ b/imagery/i.atcorr/main.cpp
> @@ -204,11 +204,11 @@ class TICache {
>  private:
>      struct RBitem set_alt_vis(double alt, double vis)
>      {
> -        struct RBitem rbitem;
> -
>          /* alt and vis must be in meters */
> -        rbitem.alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5));
> -        rbitem.vis = (int)(vis + 0.5);
> +        struct RBitem rbitem = {
> +            .alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5)),
> +            .vis = (int)(vis + 0.5),
> +        };
>  
>          return rbitem;
>      }
> ```


with this code the build failed

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp]: i did the changes it worked.
Also thanks for giving the version info.

Comment (echoix) [reply imagery/i.atcorr/main.cpp]: > (Current GRASS GIS language standard of C11 and C++11 is declared in https://github.com/OSGeo/grass/blob/main/doc/development/rfc/language_standards_support.md).

During the last community sprint, it was decided to change the checks to a minimum of c++17, and changed in https://github.com/OSGeo/grass/pull/3846

The old RFC was written in a way that it was a fixed version, and couldn't evolve. So, in the same spirit as when it was written and adopted (and the same way that we follow Python versions now), it was chosen to move the range. Adding back checks for newer language standards versions was postponed until we have Ubuntu 24.04 CI ready, in order to use newer compiler versions available (without a custom install) on the distro of users.

The C++17 was also chosen because of it aligns with one of our required dependencies, gdal, which has that standard version for as the lowest supported for their 3.9 version. That standard version only applies for our main branch, that is the 8.5 release that is planned to be out in a long time in the future.

Comment (nilason) [reply imagery/i.atcorr/main.cpp]: > > (Current GRASS GIS language standard of C11 and C++11 is declared in https://github.com/OSGeo/grass/blob/main/doc/development/rfc/language_standards_support.md).
> 
> During the last community sprint, it was decided to change the checks to a minimum of c++17, and changed in #3846
> 
> The old RFC was written in a way that it was a fixed version, and couldn't evolve. So, in the same spirit as when it was written and adopted (and the same way that we follow Python versions now), it was chosen to move the range. Adding back checks for newer language standards versions was postponed until we have Ubuntu 24.04 CI ready, in order to use newer compiler versions available (without a custom install) on the distro of users.
> 
> The C++17 was also chosen because of it aligns with one of our required dependencies, gdal, which has that standard version for as the lowest supported for their 3.9 version. That standard version only applies for our main branch, that is the 8.5 release that is planned to be out in a long time in the future.

Thanks, I was just about to file an issue to bump to C++17 support for future releases :-).

Comment (echoix) [reply imagery/i.atcorr/main.cpp]: Is it something doable to edit/upgrade the RFC instead? To make it evolvable? How long does the approval process take?

Comment (nilason) [reply imagery/i.atcorr/main.cpp]: > Is it something doable to edit/upgrade the RFC instead? To make it evolvable? How long does the approval process take?

We will probably have to make a new (superseding) RFC regarding C++ standard support. RFC 7 is voted for and accepted.

File: imagery/i.atcorr/main.cpp [hunk 2]
```diff
@@ -204,12 +204,10 @@ class TICache {
 private:
     struct RBitem set_alt_vis(double alt, double vis)
     {
-        struct RBitem rbitem;
-
+        struct RBitem rbitem = {};
         /* alt and vis must be in meters */
         rbitem.alt = (alt < 0 ? (int)(alt - 0.5) : (int)(alt + 0.5));
         rbitem.vis = (int)(vis + 0.5);
-
```
Comment (nilason) [inline imagery/i.atcorr/main.cpp:212]: Please revert this unrelated change.

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp:212]: i added the empty line back which was deleted.

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp:212]: reverting back and adding the empty line causes Clang formatting issues.

Comment (nilason) [reply imagery/i.atcorr/main.cpp:212]: > reverting back and adding the empty line causes Clang formatting issues.

It’s because of trailing whitespaces.

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp:212]: Is doing that change necessary.

Comment (nilason) [reply imagery/i.atcorr/main.cpp:212]: Not functionally, but stylistically. To be able to merge this, it has to pass all required tests.

Comment (nilason) [reply imagery/i.atcorr/main.cpp:212]: We’re back where we started with this, deleting an unrelated empty line. To revert this, add an empty line with no spaces (!) before the return line in a new commit

Comment (ShubhamDesai) [reply imagery/i.atcorr/main.cpp]: Done

**Review by nilason — APPROVED** [review]
Thanks!
